File: Python/Week 1/Weekend Project_Movie theater.py
# simulating the total revenue earned each day for a movie theater 
# import libraries
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import math


# In[2]:



def one_sample(low, high):
    """
    Generates a random sample of visitors: adults, children and seniors
    parameters
    ----------
    low: the lowest value 
    high: the highest value 
    
    
    return
    ------
    random sample of visitors
    
    """
    return np.random.randint(low=low, high=high,size=1)[0] 


def revenue(visitors, cost):
    """
    This function multiples two lists to return the total revenue for all visitors
    parameters
    ----------
    visitors: list
              number of the visitors for children, seniors and adults
              
    cost: list 
          cost of ticket for children, seniors and adults
    
    return
    ------
    total : int 
            total revenue per day
    
    """
    cost_list = np.multiply(visitors,cost)
    total = sum(cost_list)
    return total


# In[3]:


ticket_cost = [10, 15, 20]  # ticket cost child , senior , adults
visitors = []  # to store number of visitors

# data frame contening the movie
data = {
    "Movie Name": [
        "The Lion King",
        "Mad Max",
        "The Lord Of The Ring",
        "Star Wars",
        "Toy Story",
    ]
}

movies = pd.DataFrame(data, columns=["Movie Name"])

# data frame contening the days
days_name = {
    "Day": ["Sat", "Sun", "Mon", "Thues", "Wed", "Thurs", "Fri"],
    "Revenue": np.zeros(7),
}

revenue_per_day = pd.DataFrame(days_name, columns=["Day", "Revenue"])

screens = 5  # number of screens

seats = 100  # num of seats each theater has
week_days = np.zeros(7)  # list to store revenue

currentTotal = 0  # to store total revenue for each movie
total = 0  # to store revenue per day

sim = 1000  # number of simulation


# In[4]:


# simulation 1000 times
for s in range(sim):
    
    # Iterate through the week
    for i in range(len(week_days)) : 
        
        total = 0
        # Iterate through the amount of screens on a particular day
        for j in range (0,screens):

            # Generates a random sample of visitors: adults, children and seniors
            visitors_adults = one_sample(0,seats)
            visitors_child = one_sample(0, (seats - visitors_adults))
            visitors_senior = one_sample(0, (seats - (visitors_adults + visitors_child)))
           
            # store number of visitors for children, seniors and adults
            visitors = [visitors_child, visitors_senior,visitors_adults ]
            
            # calculate total revenue 
            currentTotal = revenue(visitors, ticket_cost)

            # calculate total revenue per day
            total += currentTotal 

        # store total revenue per day in the list
        week_days[i] = week_days[i] + total 


# highest summed revenue
revenue_per_day['Revenue'] = week_days

revenue_per_day


# In[8]:


# graph that shows revenue per day

y = revenue_per_day['Revenue']
low = min(y)
high = max(y)
plt.ylim([math.ceil(low-0.5*(high-low)), math.ceil(high+0.5*(high-low))])

plt.bar(revenue_per_day['Day'],revenue_per_day['Revenue'])
plt.title('Revenue per day')
plt.xlabel('Days')
plt.ylabel('Revenue $')
plt.show()


# ### Questions:
# 1. **Talk about how you refactored your code and made improvements**
# 
#     - Create functions for repeated code part and calculate revenue 
#         - `one_sample(low, high)`
#         - `revenue(visitors, cost)`
# 
#     -  Use a list instead of multiple variables. 
#         - `ticket_cost = [10, 15, 20] ` instand of `ticket_cost_child = 10, ticket_cost_senior =15, ticket_cost_adults = 20`
#         - `visitors = [visitors_child, visitors_senior,visitors_adults ]`
#     - Use DataFrame to store revenue for each day
#         - `revenue_per_day = pd.DataFrame(days_name, columns=["Day", "Revenue"])`
# 
# 
# 
# 2. **What challenges you faced**
#     - We tried differant packages and methods to generate random number, some of them required spcific input type, like `Ranmdom.Sample()` this function take a list as input. We need to pass int number as input to function, so, we used `np.random.randint`.
#     
#     
# 3. **What changes you would like to make if you had more time.**
#     - Calculating the cost based on the number of the movie view.
#     - Store the number of visitors to find out the most visitors from which category?
#     
# 

File: R/Week 1/Weekend Project_Movie Theater.R
# Team Members

# Zarah Shibli
# Fatimah Moneer Al-Rashed
# Abdulrahman Aljubaylan
# Rawan Alsaedi
########################

# Potential Questions to Answer:
# 1. Create snacks that the customers can buy and randomize who buys which snack
# 2. Pretend you own multiple theaters and run two simulations to represent each theater and plot the results
# 3. Create conditional statements for movies that may be PG-13 and children are not allowed to watch (Done)


# Cost for adults and children
ticket_cost <- 5
ticket_cost_child <- 10 

# List 5 of your favorite movies
movies <- data.frame(
  movie_name = c('Fireproof', 'Waitress', 'Twilight', 'Shooter', 'Proposal'),
  allowed_for_child = c("No","No","Yes","Yes","Yes")
  
)

screens <- 1:5 # How many screens does the theater have? (assume 1 per movie)
total_seats = 150 # How many seats does each theater hold
seats <-  (1:total_seats) # Range of seats on each theater hold
week_days <- rep(0, 7)  # Store totals for each day

visitors_adults <- data.frame() # Store number of adults visitors for each day
visitors_children <- data.frame() # Store number of children visitors for each day 

    # iterate through the week
    for (i in 1:length(week_days)){
      
      # Keep track of total revenue for the day
      
      visitors_per_day_adults = list() # Store the count watching adults  for each movie
      visitors_per_day_children = list() # Store the count watching Children  for each movie
      
      # iterate through the amount of screens on a particular day
      for (j in screens) {
        
        # Calculate how many adults and children are watching the movie
        
        # check if allowed for children to watch this movie or not
        if (movies$allowed_for_child[j]=="Yes"){
          visitors_adults_sample <- sample(seats, 1) # Return a random number for adults visitors number 
          visitors_children_sample <- sample((total_seats-visitors_adults_sample),1) # Return a random number for children visitors number 
        }
        else{
          visitors_adults_sample <- sample(seats, 1)
          visitors_children_sample = 0 # Set children visitors number by 0, because not allowed for children to watch this movie
        }
        
        # Calculate the revenue for adults and children
        revenue_adults = visitors_adults_sample * ticket_cost 
        revenue_children = visitors_children_sample * ticket_cost_child
        
        # Calculate revenue, and add to running total for the day
        total_per_day= revenue_adults + revenue_children
        
        # Additional feature: Count number of visitors for adults and children
        visitors_per_day_adults[j] <- visitors_adults_sample
        visitors_per_day_children[j] <- visitors_children_sample

      }
      # Save total to the corresponding day
      week_days[i] <- total_per_day
      
      # Store adults and children visitors number into dataframes
      visitors_adults <- rbind(visitors_adults, visitors_per_day_adults)
      visitors_children <- rbind(visitors_children, visitors_per_day_children)
    }
    
    # Rename column names
    colnames(visitors_adults) <- movies$movie_name
    colnames(visitors_children) <- movies$movie_name
    
    # View data
    View(visitors_adults)
    View(visitors_children)
   
    # Make a barchart showing total revenue per day
    # Which day had the highest revenue? 
    
    
        max_value = max(week_days) # find maximum revenue
        
        # Coloring the maximum total revenue per day
        colors <- ifelse(week_days == max_value,"green","gray")
        
        # create a bar plot fot the revenue per day
        barplot(
          week_days,
          main = "Total Revenue per Day",
          xlab = "Day",
          ylab = "Total Revenue",
          names.arg = c("Sat","Sun","Mon","Thues","Wed","Thurs","Fri"),
          col = colors # add color vector
          )
        
   
        
    # Make any other chart
        
        # Ploting number of watchers for each movie based on age (Adults or Children)
        visitors_list <- list(adults=colSums(visitors_adults),children = colSums(visitors_children))
        visitors_list
        
        par(
          mfrow=c(1,2), 
          mar=c(1,1)) # used to fix the error "figure margins to large")
        
        barplot(
          visitors_list$adults,
          main = "Total of Adults watchers for each movie",
          xlab = "Movies",
          ylab = "Number of watchers",
          col = c("red")
        )
        
        barplot(
          visitors_list$children,
          main = "Total of children watchers for each movie ",
          xlab = "Movies",  
          ylab = "Number of watchers",
          col = c("green")
        )
        